enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,1202 @@
+                    "url": "registry.ci.openshift.org/..."
+                },
+
+                # If the component was upgraded from known versions.
+                "from": [
+                    {
+                        "name": "...",
+                        "source": "..."
+                    }
+                ]
+            },
+            "openshift:hosted:cluster": {
+                "component": {
+                    "product": "openshift",
+                    "type": "hosted",
+                },
+
+                "version": "4.17.0",
+
+                "source": {
+                    "type": "image",
+                    "digest": "sha256:...",
+                    "url": "registry.ci.openshift.org/..."
+                },
+
+                "from": [
+                    {
+                        "name": "...",
+                        "source": "..."
+                    }
+                ]
+            },
+          
+            # origin will aggregate versions advertised by all 
+            # components.
+            "openshift:payload:hyperkube": {
+                # ...
+            },
+        },
+    },
+
+    # Information about how the test was run which should
+    # not alter the outcome of the test.
+    "context": {
+        # A small number of seeds can be used to randomize test
+        # parallelization and bucketing. Analyzing seeds against
+        # one another should highlight parallelization issues.
+        # Using the same seed with the same tests should result
+        # in identical planning and parallelization. 
+        "seed": 15,
+
+        # A sha256 of the tests IDs which were run. This, plus
+        # seed should ensure identical execution pattern between
+        # two runs.
+        "testHash": "deadbeef",
+
+        # In the future, we may run tests several times against
+        # the same cluster in order to aggregate
+        # test results more quickly. We may want to 
+        # analyze results including this dimension if, for example,
+        # we want to check whether the first run is as consistent
+        # as the following N.
+        "run": 0,
+
+        "job": {
+            # Name of the job which drove the test
+            "name": "...",
+            # Where the job information can be reviewed.
+            "url": "..."
+        },
+
+        # Attributes of the host running the tests (not the 
+        # system under test) or cloud account used. This can 
+        # help us statistically analyze whether one system 
+        # involved in running the testing is introducing 
+        # failures that another is not (e.g. because it has 
+        # a flakey network).
+        "testDriver": {
+            "cluster": "build01",
+            "arch": "arm64",
+            # Cloud account in which the ephemeral cluster was 
+            # installed.
+            "account": "34908092383"
+        }
+    }
+}
+```
+
+### Risks and Mitigations

```

Comment:
With all tests in `openshift/origin`, we need to bump k8s.io/kubernetes with many useful e2e functions only there. When we move the tests to individual repos, all these repos will endure pain with updating k8s.io/kubernetes. Upstream does not even pretend it has a stable API and it often breaks.

Alternatively, we would need to spend some non-trivial time rewriting the tests not to depend on k8s.io/kubernetes/test/e2e/framework.

Reply:
Would it be helpful to have a discussion about your particular use case? I might be missing some context. What e2e functions are you using from k8s.io/kubernetes/test/e2e/framework? Which tests are you looking to move to external binaries for which repos?

- Our next OTE customer after migrating k8s-tests-ext is ovn-kubernetes, and the ginkgo tests already exist in their repos to run them.  We'd mostly be running them unmodified from upstream

-  The other use case we looked at was moving QE's openshift-tests-private to the component repos, and those don't use anything from k8s.io/kubernetes/test/e2e/framework.

Reply:
I understand the proposal that tests from origin should eventually move to  individual team repos. If I look what imports k/k in today's origin: ```
$ grep -R -l k8s.io/kubernetes/test/e2e/framework  test/extended  | wc -l
194
```

If we move these 194 files to individual team repos, these teams will either need to use some other test framwork than `k8s.io/kubernetes/test/e2e` or rebase unstable k/k regularly.

If the tests can stay in origin, then I'm fine.

Reply:
There's no plan to plan to force anyone to move, but I'm also not entirely sure I understand the problem.  The vendoring of k8s.io/kubernetes/test/e2e/framework would just be in the component repo - there's no interdependency between origin's deps and an OTE binary's deps.